Implement RT priority assignment

balandbal opened the pull request:
The [README](https://github.com/ros-industrial/kuka_experimental/tree/melodic-devel/kuka_rsi_hw_interface/krl/KR_C4) for the (KRC4) RSI hardware interface recommends setting the PC real-time priority for it if the connection to RSI is sporadically dropped. The hardware interface for Universal Robots arms [sets this priority when it can](https://github.com/UniversalRobots/Universal_Robots_ROS_Driver/blob/395c0541b20d0da2cd480e2ad85b2100410fb043/ur_robot_driver/src/hardware_interface_node.cpp#L63); I thought it sensible to adopt that part of the code into the KUKA interface.

As far as I could determine at a glance, the implementation should not conflict with anything already in the KUKA interface, though my experience with C++ and the source code of the hardware interface is limited.

Once our cell is built completely in the following weeks, I will be using the patch more extensively. If there are any specific tests that would be great to run, I am willing to try and set them up.

simonschmeisser commented on kuka_rsi_hw_interface/src/kuka_hardware_interface_node.cpp:
```diff
@@ -50,6 +50,58 @@ int main(int argc, char** argv)
+      ret = pthread_getschedparam(this_thread, &policy, &params);
+      if (ret != 0)
+      {
+        std::cout << "Couldn't retrieve real-time scheduling paramers" << std::endl;
```
Shouldn't this be `ROS_ERROR_STREAM` as well?

balandbal replied:
Yes. Adjusted in [06f0ab7](https://github.com/ros-industrial/kuka_experimental/pull/257/commits/06f0ab7071e101ba005ab343d92d1439cdadf8a2)

simonschmeisser commented on kuka_rsi_hw_interface/src/kuka_hardware_interface_node.cpp:
```diff
@@ -50,6 +50,58 @@ int main(int argc, char** argv)
+      ret = pthread_getschedparam(this_thread, &policy, &params);
+      if (ret != 0)
+      {
+        ROS_ERROR_STREAM("Couldn't retrieve real-time scheduling paramers");
```
```suggestion
        ROS_ERROR_STREAM("Couldn't retrieve real-time scheduling parameters");
```
copy paste of a typo present in UR as well

balandbal replied:
Thanks for catching it; taken care of in [6eb8549](https://github.com/ros-industrial/kuka_experimental/pull/257/commits/6eb85497e0f8a4d654cc285cc8b5a4bec544e850)

simonschmeisser commented:
Thanks for your PR! This looks fine now from the formal side. I have no experience with RSI (yet) so I would like to wait for either you or me to test this patch before merging and ideally for someone with real time experience to also review it.

gavanderhoorn commented:
I believe, as this was copied from the/a UR driver, that the provenance of this code should be clearly stated, including the license it is covered by, and the original author(s).

IIRC, the code actually comes from the Franka Emika driver, so perhaps the UR driver authors already didn't abide by the license(s) there, but that doesn't mean we can do the same here.

balandbal commented:
> I believe, as this was copied from the/a UR driver, that the provenance of this code should be clearly stated, including the license it is covered by, and the original author(s).
> 
> IIRC, the code actually comes from the Franka Emika driver, so perhaps the UR driver authors already didn't abide by the license(s) there, but that doesn't mean we can do the same here.

I added the license, copyright and author information in [c51e1b4](https://github.com/ros-industrial/kuka_experimental/pull/257/commits/c51e1b40471d4f70b9fcb1a415d9509295cb4a7b).

- To make the distinction cleaner between the BSD and Apache-2.0 code (the UR driver comes subject to that), I factored out the priority assigning into the separate header [kuka_rsi_hw_interface/assign_rt_priority.h](https://github.com/balandbal/kuka_experimental/blob/assign_rt_priority/kuka_rsi_hw_interface/include/kuka_rsi_hw_interface/assign_rt_priority.h). 
- Apache-2.0 requires retaining a copy of the license in the project root; kuka_experimental already has one, so I thought it was enough not to add that again.
- To notify those reading the source about the mixed licensing, I modified the Software License Agreements in [kuka_hardware_interface_node.cpp](https://github.com/balandbal/kuka_experimental/blob/assign_rt_priority/kuka_rsi_hw_interface/src/kuka_hardware_interface_node.cpp) and [kuka_hardware_interface.h](https://github.com/balandbal/kuka_experimental/blob/assign_rt_priority/kuka_rsi_hw_interface/include/kuka_rsi_hw_interface/kuka_hardware_interface.h).
- To credit the original author, I added them to the [package.xml](https://github.com/balandbal/kuka_experimental/blob/c51e1b40471d4f70b9fcb1a415d9509295cb4a7b/kuka_rsi_hw_interface/package.xml#L7) and as authors to the [new header](https://github.com/balandbal/kuka_experimental/blob/c51e1b40471d4f70b9fcb1a415d9509295cb4a7b/kuka_rsi_hw_interface/include/kuka_rsi_hw_interface/assign_rt_priority.h#L27).

Adding a notice about the mixed licensing seemed excessive for [kuka_hardware_interface.cpp](https://github.com/balandbal/kuka_experimental/blob/assign_rt_priority/kuka_rsi_hw_interface/src/kuka_hardware_interface.cpp), though it also includes the header with the priority assignment through [kuka_hardware_interface.h](https://github.com/balandbal/kuka_experimental/blob/assign_rt_priority/kuka_rsi_hw_interface/include/kuka_rsi_hw_interface/kuka_hardware_interface.h).

gavanderhoorn commented:
thanks for the update.

However, calling a function does not change the copyright on the file that calls it. Neither does including a `.h`. So the changes to the files that just call the new function or include the header would not be needed (`kuka_rsi_hw_interface/include/kuka_rsi_hw_interface/kuka_hardware_interface.h` fi).

I'm also not sure whether Felix would have to be added to the manifest: *many* people have contributed to these packages, and we don't list all of them. Only the main contributors are listed.
